src/vdaf/hits.rs

```diff
@@ -0,0 +1,749 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! **(NOTE: This module is experimental. Applications should not use it yet.)** This module
+//! implements the VDAF [[VDAF]] used for the privacy-preserving heavy hitters protocol of
+//! [[BBCG+21]]. It is constructed from an Incremental Distributed Point Function (IDPF).
+//!
+//! [VDAF]: https://cjpatton.github.io/vdaf/draft-patton-cfrg-vdaf.html
+//! [BBCG+21]: https://eprint.iacr.org/2021/017
+
+use std::array::IntoIter;
+use std::collections::HashSet;
+use std::convert::TryFrom;
+
+use crate::field::{split_vector, FieldElement};
+use crate::fp::log2;
+use crate::prng::Prng;
+use crate::vdaf::suite::{Key, KeyDeriver, KeyStream, Suite};
+use crate::vdaf::{Share, VdafError};
+
+/// An input for an IDPF ([`Idpf`]).
+#[derive(Clone, Debug, Hash, Eq)]
+pub struct IdpfInput<'a> {
+    data: &'a [u8],
+    len: usize,
+}
+
+impl<'a> IdpfInput<'a> {
+    /// Constructs an IDPF input using the first `len` bits of `data`.
+    pub fn new(data: &'a [u8], len: usize) -> Result<Self, VdafError> {
+        if len > data.len() << 3 {
+            return Err(VdafError::Uncategorized(format!(
+                "desired bit length ({} bits) exceeds data length ({} bytes)",
+                len,
+                data.len()
+            )));
+        }
+
+        Ok(Self { data, len })
+    }
+
+    /// Return the `bit`-th bit of the input. Bounds checking is performed by the caller.
+    fn bit(&self, bit: usize) -> usize {
+        let i = bit >> 3;
+        let j = ((i << 3) ^ bit) as u8;
+        (self.data[i] >> j) as usize & 1
+    }
+
+    /// Return the index of input `self` in the look-up table of a `ToyIdpf` instance.
+    fn index(&self) -> usize {
+        let mut index = 1 << self.len;
+        for i in 0..self.len {
+            index += self.bit(i) << i;
+        }
+        index - 1
+    }
+
+    /// Construct a new input that is a prefix of `self`. Bounds checking is performed by the
+    /// caller.
+    fn prefix(&self, len: usize) -> Self {
+        Self {
+            data: self.data,
+            len,
+        }
+    }
+}
+
+impl<'a> PartialEq<IdpfInput<'a>> for IdpfInput<'a> {
```

Comment:
We provide a `PartialEq` that only considers the first `self.len` bytes of `self.data`. We should provide an implementation of `Hash` that works the same way, or data structures will behave surprisingly!

Reply:
On second thought, the derived `Hash` impl is sufficient. Given `IdpfInput::new(b"hello", 40)` and `IdpfInput::new(b"hello", 39)`, the hash will include the entire string "hello" in both cases, but the different `len` values will cause the resulting hash to differ, satisfying the requirements of the relationship between `Hash` and `Eq`.